### PR TITLE
Add FileHelper::isEmptyDirectory() and improve FileHelper::unlink()

### PR DESCRIPTION
**Pull request by vjik**
| Q             | A
| ------------- | ---
| Is bugfix?    | ✔️
| New feature?  | ✔️
| Breaks BC?    | ❌
| Fixed issues  | #21

### CODE REVIEW

File: src/FileHelper.php [hunk 1]
```diff
@@ -213,11 +214,51 @@ public static function unlink(string $path): bool
...
+            return unlink($path);
+        } catch (Exception $e) {
+            // Last resort measure for Windows: remove via command DEL.
+            // We can try remove only file, because when remove directory command DEL removed
+            // only files, directory not removed.
+            // @see https://docs.microsoft.com/en-us/windows-server/administration/windows-commands/del
+            if (is_dir($path) && !static::isEmptyDirectory($path)) {
+                return false;
+            }
+
```
Comment (roxblnfk) [inline src/FileHelper.php:236]:
```suggestion
```

File: src/FileHelper.php [hunk 2]
```diff
@@ -213,11 +214,51 @@ public static function unlink(string $path): bool
...
         }
 
+        return static::windowsUnlink($path);
+    }
+
+    private static function windowsUnlink(string $path): bool
+    {
         if (is_link($path) && is_dir($path)) {
             return rmdir($path);
         }
```
Comment (roxblnfk) [inline src/FileHelper.php:224]:
```suggestion
        if (is_dir($path)) {
            if (is_link($path) || self::isEmptyDirectory($path)) {
                return rmdir($path);
            } else {
                return false;
            }
        }
```

**Comment by samdark** [comment]
Here's a simpler fix: https://github.com/yiisoft/files/pull/25

File: src/FileHelper.php (hunk 2 above)
Comment (vjik) [reply src/FileHelper.php:218]: Refactored windowsUnlink as https://github.com/yiisoft/files/pull/25

File: src/FileHelper.php [hunk 3]
```diff
@@ -207,19 +206,42 @@ public static function clearDirectory(string $directory, array $options = []): v
...
 
+        if (!is_writable($path)) {
+            chmod($path, 0777);
+        }
+
         return unlink($path);
     }
 
+    public static function isEmptyDirectory(string $path): bool
```
Comment (samdark) [inline src/FileHelper.php:227]: What's the use case for it?

File: src/FileHelper.php [hunk 4]
```diff
@@ -207,19 +206,42 @@ public static function clearDirectory(string $directory, array $options = []): v
...
+        $handle = static::openDirectory($path);
+        while (($file = readdir($handle)) !== false) {
+            if ($file != '.' && $file != '..') {
+                closedir(@$handle);
+                return false;
+            }
+        }
+        closedir($handle);
+
+        return true;
```
Comment (samdark) [inline src/FileHelper.php:242]:
```suggestion
        if (!is_dir($path)) {
            return false;
        }
        return !(new \FilesystemIterator($path))->valid();
```

Comment (samdark) [reply src/FileHelper.php:242]: Not sure about efficiency but it's way less code.

File: src/FileHelper.php [hunk 5]
```diff
@@ -207,19 +206,42 @@ public static function clearDirectory(string $directory, array $options = []): v
      */
     public static function unlink(string $path): bool
     {
-        $isWindows = DIRECTORY_SEPARATOR === '\\';
-
-        if (!$isWindows) {
-            return unlink($path);
-        }
+        return DIRECTORY_SEPARATOR === '\\' // is windows
+            ? static::windowsUnlink($path)
```
Comment (samdark) [inline src/FileHelper.php:210]: I don't think it was bad in a single method.

**Comment by samdark** [comment]
Let's do the following:

1. Merge `windowsUnlink()` with `unlink()`. There's no need to have 2 methods and original one is clear enough.
2. Add phpdoc for `isEmptyDirectory()`.

**Comment by vjik** [comment]
> Let's do the following:
> 
> 1. Merge `windowsUnlink()` with `unlink()`. There's no need to have 2 methods and original one is clear enough.
> 2. Add phpdoc for `isEmptyDirectory()`.

Done

**Review by samdark — APPROVED** [review]

**Pull request merged**